File: src/main/java/de/schramm/royalbash/controller/service/gameevent/TurnEndedEvent.java
package de.schramm.royalbash.controller.service.gameevent;

import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class TurnEndedEvent implements GameEvent {

    private final String playerId;

    @Override
    public Game invoke(Game game) {
        return game.findPlayer(playerId)
                .filter(player -> game.getPlayerOnTurn().equals(player))
                .map(player -> game.toBuilder()
                    .playerOnTurn(player.equals(game.getPlayer1())
                            ? game.getPlayer2()
                            : game.getPlayer1())
                    .build())
                .orElse(game);
    }
}

File: src/main/kotlin/de/schramm/royalbash/controller/service/gameevent/TurnEndedEvent.kt
package de.schramm.royalbash.controller.service.gameevent

import de.schramm.royalbash.controller.service.core.Game

data class TurnEndedEvent(val playerId: String) : GameEvent {

    constructor(): this(playerId = "")

    override fun invoke(game: Game): Game {
        return game.findPlayer(playerId)
                .filter { player -> game.playerOnTurn == player }
                .map { player -> game.copy(playerOnTurn = if (player == game.player1) game.player2 else game.player1) }
                .orElse(game)
    }
}
